File: tennis-intelligence-platform-web/src/types/researchDashboard.ts
// types/researchDashboard.ts — matches api/schemas/research_dashboard.py exactly.

export interface MetricWithCI {
  point_estimate: number;
  ci_lower: number;
  ci_upper: number;
}

export interface ReliabilityPoint {
  bin_index: number;
  n: number;
  mean_predicted: number;
  observed_win_rate: number;
  calibration_gap: number;
}

export interface PredictionHistogram {
  bin_edges: number[];
  counts: number[];
}

export interface EngineResearchStats {
  display_name: string;
  n_points: number;
  log_loss: MetricWithCI;
  brier: MetricWithCI;
  ece: number;
  sharpness: number;
  reliability_diagram: ReliabilityPoint[];
  prediction_histogram: PredictionHistogram;
}

export interface ResearchDashboardResponse {
  n_matches: number;
  n_points: number;
  holdout_year: number;
  is_full_holdout: boolean;
  n_calibration_bins: number;
  n_bootstrap: number;
  engines: Record<string, EngineResearchStats>;
}

File: tennis-intelligence-platform/api/schemas/research_dashboard.py
"""
schemas/research_dashboard.py — Pydantic response model for the Research Dashboard.
Field names match tennis_intel.serving.research_dashboard_service's own JSON export
shape exactly.
"""

from __future__ import annotations

from pydantic import BaseModel


class MetricWithCI(BaseModel):
    point_estimate: float
    ci_lower: float
    ci_upper: float


class ReliabilityPoint(BaseModel):
    bin_index: int
    n: int
    mean_predicted: float
    observed_win_rate: float
    calibration_gap: float


class PredictionHistogram(BaseModel):
    bin_edges: list[float]
    counts: list[int]


class EngineResearchStats(BaseModel):
    display_name: str
    n_points: int
    log_loss: MetricWithCI
    brier: MetricWithCI
    ece: float
    sharpness: float
    reliability_diagram: list[ReliabilityPoint]
    prediction_histogram: PredictionHistogram


class ResearchDashboardResponse(BaseModel):
    n_matches: int
    n_points: int
    holdout_year: int
    is_full_holdout: bool
    n_calibration_bins: int
    n_bootstrap: int
    engines: dict[str, EngineResearchStats]